types/d3-collection/index.d.ts

```diff
@@ -16,106 +17,463 @@ export interface Stringifiable {
+    /**
+     * Returns an array of values for every entry in this map.
+     * The order of the returned values is arbitrary.
+     */
     values(): T[];
+    /**
+     * Returns an array of key-value objects for each entry in this map. The order of the returned entries is arbitrary.
+     * Each entry’s key is a string, but the value can have arbitrary type.
+     */
     entries(): Array<{ key: string, value: T }>;
+    /**
+     * Calls the specified function for each entry in this map and returns undefined.
+     * The iteration order is arbitrary.
+     *
+     * @param func Function to call for each entry. The function is passed the entry’s value and key as arguments,
+     * followed by the map itself.
+     */
     each(func: (value: T, key: string, map: Map<T>) => void): void;
+    /**
+     * Returns true if and only if this map has zero entries.
+     */
     empty(): boolean;
+    /**
+     * Returns the number of entries in this map.
+     */
     size(): number;
 }
 
+/**
+ * Constructs a new empty map.
+ *
+ * The generic refers to the data type of the map entry values.
+ */
 export function map<T>(): Map<T>;
+/**
+ * Constructs a new map by copying another map.
+ *
+ * The generic refers to the data type of the map entry values.
+ *
+ * @param d3Map A D3 Map.
+ */
 export function map<T>(d3Map: Map<T>): Map<T>;
+/**
+ * Constructs a new map by copying all enumerable properties from the specified object into this map.
+ *
+ * The generic refers to the data type of the map entry values.
+ *
+ * @param object Object to construct the map from.
+ */
 export function map<T>(object: { [key: string]: T }): Map<T>;
+/**
+ * Constructs a new map by copying all enumerable properties from the specified object into this map.
+ *
+ * The generic refers to the data type of the map entry values.
+ *
+ * @param object Object to construct the map from.
+ */
 export function map<T>(object: { [key: number]: T }): Map<T>;
-export function map<T>(array: T[], key?: (value: T, i?: number, array?: T[]) => string): Map<T>;
-export function map(object: any): Map<any>; // TODO: When upgrading definitions to use TS 2.2+, use "object" data type for argument
+/**
+ * Constructs a new map from the elements of an array.
+ * An optional key function may be specified to compute the key for each value in the array.
+ *
+ * The generic refers to the data type of the map entry values.
+ *
+ * @param array Array to convert into a map
+ * @param key An optional key function. The functions is invoked for each element in the array being passed
+ * the element's value , it's zero-based index in the array, and the array itself. The function must return a unique string
+ * to be used as the map entry's key.
+ */
+export function map<T>(array: T[], key?: (value: T, i: number, array: T[]) => string): Map<T>;
+/**
+ * Constructs a new map by copying all enumerable properties from the specified object into this map.
+ */
+export function map<T = any>(object: object): Map<T>;
```

Comment:
Same `keyof` suggestion applies here.

Reply:
Good old `keyof`. Yeah we had put that on hold back in the day, because the release was too fresh. I don't disagree with your point that the two code pieces you called out where the `object` type is used, are a little sub-optimal from a type-safety perspective.

I just played around with it a little in the context of d3-collection (`keys(...)`, `values(...)`, `entries(...)` and `map(...)`. A couple of things stand out:

* the functionality in d3-collection hits only enumerable properties (including those on `prototype`), `keyof` on the other hand is not restricted to `enumerable`s. So, e.g. calling `keyof A` on a `type A = Array<number | string>` will be vastly different and so will be `A[keyof A]`. Similarly, if one used instances of a class like:
```ts
class TestClass {
    a: 'test';
    b: 123;
    c: 'type';
    private d: 'hidden';
    test() { }
}
```
`keyof ` ignores the private member `d`, but includes the public `test()` method.

* A change management consideration is that so far in all overloaded signatures the generics referred to the data type of the value of the properties. Using the `TObject`-approach diverts from that cognitive model. If we additionally drop the signatures with `xyz<T>(object: { [key: string]: T })` arguments, we might break folks how explicitly set the generic `T`

More to come...

Reply:
...to continue:
So, in short, I am wondering, if using `keyof` in the case of d3-collection does more harm than good? If we don't use it, at a minimum we would have to keep the overloads for key/value object such as:
```ts
export function values<T>(object: { [key: string]: T }): T[];
```
_prior_ to the final `obj: object` signature, the compiler picks it up first and prevents the "oops" you mentioned for your `dict` example. The key/value signature also will infer a `values(TestClass)` or `values(document)` return type of `any[]` (less specific, but correct).

What the `key/value` signature won't handle alone, are `array`s or more exotic exercises, such as [this actual but somewhat artificial d3-collection test case](https://github.com/d3/d3-collection/blob/1eff7233bac174045331d275e7e0cb793dfdaa0d/test/values-test.js#L11).

Reply:
@gustavderdrache Correction on my part, the key/value signatures will only be constraining for the `dict` example, if we leave the `object` signature as:
```ts
function values(obj: object): any[];
```